gadget/raw.go

```diff
@@ -0,0 +1,295 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package gadget
+
+import (
+	"bytes"
+	"crypto"
+	_ "crypto/sha1"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/osutil"
+)
+
+// RawStructureWriter implements support for writing raw (bare) structures.
+type RawStructureWriter struct {
+	rootDir string
+	ps      *PositionedStructure
+}
+
+// NewRawStructureWriter returns a writer for given structure, that will load
+// the structure content data from provided root directory.
+func NewRawStructureWriter(rootDir string, ps *PositionedStructure) (*RawStructureWriter, error) {
+	if !ps.IsBare() {
+		return nil, fmt.Errorf("structure %v is not bare", ps)
+	}
+	rw := &RawStructureWriter{
+		rootDir: rootDir,
+		ps:      ps,
+	}
+	return rw, nil
+}
+
+// writeRawStream writes the input stream in that corresponds to provided
+// positioned content. The number of bytes read from input stream must match
+// exactly the declared size of positioned content entry.
+func writeRawStream(out io.WriteSeeker, pc *PositionedContent, in io.Reader) error {
+	if _, err := out.Seek(int64(pc.StartOffset), io.SeekStart); err != nil {
+		return fmt.Errorf("cannot seek to content start offset 0x%x: %v", pc.StartOffset, err)
+	}
+
+	_, err := io.CopyN(out, in, int64(pc.Size))
+	if err != nil {
+		return fmt.Errorf("cannot write image: %v", err)
+	}
+	return nil
+}
+
+// writeRawImage writes a single image described by a positioned content entry.
+func (r *RawStructureWriter) writeRawImage(out io.WriteSeeker, pc *PositionedContent) error {
+	if pc.Image == "" {
+		return fmt.Errorf("internal error: no image defined")
+	}
+	img, err := os.Open(filepath.Join(r.rootDir, pc.Image))
+	if err != nil {
+		return fmt.Errorf("cannot open image file: %v", err)
+	}
+	defer img.Close()
+
+	return writeRawStream(out, pc, img)
+}
+
+// Write will write whole contents of a structure into the output stream.
+func (r *RawStructureWriter) Write(out io.WriteSeeker) error {
+	for _, pc := range r.ps.PositionedContent {
+		if err := r.writeRawImage(out, &pc); err != nil {
+			return fmt.Errorf("failed to write image %v: %v", pc, err)
```

Comment:
would the index of the failed pc (the first arg of the range i mean) be helpful in the error?

Reply:
Positioned content String() already does that and includes the actual index from gadget yaml.